### ISSUE TITLE
Port `twisted.internet.test.reactormixins` to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5967]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5967|
|Type|enhancement|
|Created|2012-09-12 13:50:39Z|
|Branch|https://github.com/twisted/twisted/tree/reactormixins-python3-5967|

This module is the basis of a decent sized chunk of the test suite for the reactor implementations.  Porting it to Python 3 is a necessary precondition to being able to run that part of the test suite on Python 3, and therefore a precondition for being able to port the reactors to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__5967 5967
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_reactormixins_python3_5967 branches/reactormixins-python3-5967
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1347457839000000 1347457839000000
changetime__1347475606000000 1347475606000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [35593]) Branching to 'reactormixins-python3-5967'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @itamarst|
|-|-|

Moved some code around.  Adjusted some imports.  Removed use of `assertFailure` from `test_unix.py`, to work with change from `TestCase` to `SynchronousTestCase` in `reactormixins.py`.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/reactormixins-python3-5967)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [35606]) Switch another instance of assertFailure to manual Failure checking

refs [#5967](https://github.com/twisted/twisted/issues/5967)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. filter() is lazy in Python 3, it seems.
  1. Given the above wouldn't have been caught by anything, we should open a ticket for writing tests for ReactorBuilder. Especially since I want to expose some of this infrastructure, to allow users to write tests that have a reactor per-test (unless we have a full fake reactor that'll be ready soon).
  1. Given we've both implemented variations, possibly SyncTestCase should have a sync version of assertFailure? Or maybe the use cases are sufficiently diverse that there's no point.
  1. Add future imports.

Please fix, and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [35611]) Avoid lazy filter.  Also fix `__all__` definition.

refs [#5967](https://github.com/twisted/twisted/issues/5967)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [35614]) Merge reactormixins-python3-5967

Author: exarkun
Reviewer: itamar
Fixes: [#5967](https://github.com/twisted/twisted/issues/5967)

Port `twisted.internet.test.reactormixins` to Python 3, in anticipation of porting some tests
based on it to Python 3.
